### PR TITLE
max_memory_usage_for_session setting [currenty on hold]

### PR DESCRIPTION
**Pull request by myrrc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Another setting to control memory consumption, works similar to `max_memory_usage_for_user`, but limits the RAM amount for a single session. The default value (as proposed) is 0.

Detailed description / Documentation draft:

See docs (`operations/settings/query-complexity/#settings_max_memory_usage_for_session`).

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -306,6 +306,7 @@ struct Settings : public SettingsCollection<Settings>
     \
     M(SettingUInt64, max_memory_usage, 0, "Maximum memory usage for processing of single query. Zero means unlimited.", 0) \
     M(SettingUInt64, max_memory_usage_for_user, 0, "Maximum memory usage for processing all concurrently running queries for the user. Zero means unlimited.", 0) \
+    M(SettingUInt64, max_memory_usage_for_session, 0, "Maximum memory usage for processing all concurrently running queries for a user session. Zero means unlimited.", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:309]: It's unclear what is the motivation of this setting.
Is it for Memory and Temporary tables?

Comment (alex-zaitsev) [reply src/Core/Settings.h:309]: Yes, this is required when there are multiple temp tables involved, so sessions can easily get out of user and query limits.

**Comment by myrrc** [comment]
Proposed changes:
Currently in `Interpreters/ProcessList.h` we have the hierarchy 
```
ProcessList =>
    (Username => RAM usage info);
    (Username => ProcessListForUser)
```
the latter responsible for tracking user memory (and limiting it by the `max_memory_usage_for_user` setting). 
As we introduce another level of abstraction (memory consumption setting for a session) that is located somewhat between the query and the user (user can have multiple sessions each holding multiple queries), I suggest modifying the hierarchy this way:
```
ProcessList =>
    (Session id => RAM usage info, session-local);
    (Session id) =>
        (Username => RAM usage info, user-local);
        (Username => ProcessListForUser);
```
So the tracker inside the first container (`session_id` to `users`) would aggregate and control the per-session memory consumption, while the `user` => `queries`, as currently implemented, would control per-user memory.

**Comment by alexey-milovidov** [comment]
Ok.

But TCPHandler don't have session ids.

**Comment by agelwarg** [comment]
We just hit this today and it brought down 2 production nodes.  We have a particular application path triggered by our users which creates temporary tables that, in aggregate, lead to an OOM on our Clickhouse nodes.  We recognize that this is bad application code and we are investigating how to address it.  However, this feature would provide somewhat of a safety net such that even though this one path may fail, it wouldn't bring down the entire node.

Are there still open questions in terms of how this is being implemented that our use case can help with?

**Comment by myrrc** [comment]
> We just hit this today and it brought down 2 production nodes. We have a particular application path triggered by our users which creates temporary tables that, in aggregate, lead to an OOM on our Clickhouse nodes. We recognize that this is bad application code and we are investigating how to address it. However, this feature would provide somewhat of a safety net such that even though this one path may fail, it wouldn't bring down the entire node.
> 
> Are there still open questions in terms of how this is being implemented that our use case can help with?

Hi agelwarg!
I have some ideas how to do the ticket but currently is on hold for several reasons:

- It's quite tricky. Modifying the `ProcessList` struct should be done with great care as it's used by `Context`, which appears to be error-prone and hard to understand. Moreover, it's somewhat a vital header so it would trigger a several hours recompilation with an each code addition.
- I have a more important task (related to the ZooKeeper), which would take about 20 days. 

Maybe you should contact someone from the Yandex team or @alex-zaitsev and further discuss this task's priority with them.

Have a good day, 
Mike.

**Pull request closed without merging**